File: src/paymentservice/jwt_compression.js
// JWT Compression Library for Node.js
// 3-header design: header + payload + signature for IdP compatibility

const logger = require('./logger');

// Note: JWT header is always transmitted via x-jwt-header
// No default header - supports all IdPs (Auth0, Okta, Azure, Google with kid/jku/x5t)

/**
 * Check if JWT compression is enabled via environment variable
 */
function isJWTCompressionEnabled() {
  return process.env.ENABLE_JWT_COMPRESSION === 'true';
}

/**
 * Decompose JWT for optimized transmission
 * @param {string} jwt - Full JWT token
 * @returns {Object} Object with header, payload (raw JSON), and signature
 * Operations: 1 base64 decode (payload only)
 */
function decomposeJWT(jwt) {
  if (!jwt) {
    return null;
  }

  const parts = jwt.split('.');
  if (parts.length !== 3) {
    logger.warn('Invalid JWT format');
    return null;
  }

  const [headerB64, payloadB64, signatureB64] = parts;

  try {
    // Decode payload only - ONLY OPERATION
    const payloadJson = Buffer.from(payloadB64, 'base64url').toString('utf8');

    const result = {
      header: headerB64,       // Keep as-is (supports IdPs with kid, jku, etc.)
      payload: payloadJson,    // Raw JSON (~25% smaller than base64)
      signature: signatureB64  // Keep as-is
    };

    return result;
  } catch (err) {
    logger.warn(`Failed to decompose JWT: ${err.message}`);
    return null;
  }
}

/**
 * Reassemble JWT from header, raw JSON payload, and signature
 * @param {Object} metadata - gRPC metadata object containing x-jwt-header, x-jwt-payload, and x-jwt-sig headers
 * @returns {string|null} Reassembled JWT or null
 * Operations: 1 base64 encode (payload only)
 */
function reassembleJWT(metadata) {
  // Check for compressed JWT components (new format)
  const payloadHeader = getMetadataValue(metadata, 'x-jwt-payload');
  const signature = getMetadataValue(metadata, 'x-jwt-sig');

  const headerB64 = getMetadataValue(metadata, 'x-jwt-header');
  
  if (payloadHeader && signature && headerB64) {
    try {
      // Base64url encode the raw JSON payload - ONLY OPERATION
      const payloadB64 = Buffer.from(payloadHeader, 'utf8').toString('base64url');

      // Reassemble JWT using original header
      const jwt = `${headerB64}.${payloadB64}.${signature}`;

      return jwt;
    } catch (err) {
      logger.warn(`Failed to reassemble JWT: ${err.message}`);
      return null;
    }
  }

  // Fall back to standard authorization header
  const authHeader = getMetadataValue(metadata, 'authorization');
  if (authHeader && authHeader.startsWith('Bearer ')) {
    const jwt = authHeader.substring(7);
    logger.debug(`JWT extracted from authorization header (${jwt.length} bytes)`);
    return jwt;
  }

  return null;
}

/**
 * Get metadata value (handles gRPC metadata format)
 * @param {Object} metadata - gRPC metadata object
 * @param {string} key - Metadata key
 * @returns {string|null} Metadata value or null
 */
function getMetadataValue(metadata, key) {
  try {
    const value = metadata.get(key);
    
    if (value && value.length > 0) {
      // Handle both string and Buffer values
      const val = value[0];
      if (Buffer.isBuffer(val)) {
        return val.toString('utf8');
      }
      return val;
    }
  } catch (err) {
    logger.warn(`Error getting metadata '${key}': ${err.message}`);
  }
  return null;
}

/**
 * Add compressed JWT to metadata
 * @param {Object} metadata - gRPC metadata object
 * @param {string} jwt - Full JWT token
 */
function addCompressedJWT(metadata, jwt) {
  if (!jwt || !isJWTCompressionEnabled()) {
    // Fallback to standard authorization header
    metadata.set('authorization', `Bearer ${jwt}`);
    return;
  }

  const components = decomposeJWT(jwt);
  if (!components) {
    // Failed to decompose, use standard header
    metadata.set('authorization', `Bearer ${jwt}`);
    return;
  }

  // Add compressed components: header + raw JSON payload + signature
  metadata.set('x-jwt-header', components.header);
  metadata.set('x-jwt-payload', components.payload);
  metadata.set('x-jwt-sig', components.signature);
}

module.exports = {
  isJWTCompressionEnabled,
  decomposeJWT,
  reassembleJWT,
  addCompressedJWT,
  getMetadataValue
};

File: src/emailservice/jwt_compression.py
"""JWT Compression Library for Python
3-header design: header + payload + signature for IdP compatibility
"""

import json
import base64
import os
import logging

logger = logging.getLogger(__name__)

# Note: JWT header is always transmitted via x-jwt-header
# No default header - supports all IdPs (Auth0, Okta, Azure, Google with kid/jku/x5t)


def is_jwt_compression_enabled():
    """Check if JWT compression is enabled via environment variable"""
    return os.environ.get('ENABLE_JWT_COMPRESSION', 'false').lower() == 'true'


def base64url_decode(data):
    """Decode base64url string"""
    # Add padding if needed
    missing_padding = len(data) % 4
    if missing_padding:
        data += '=' * (4 - missing_padding)
    return base64.urlsafe_b64decode(data)


def base64url_encode(data):
    """Encode to base64url string (no padding)"""
    if isinstance(data, str):
        data = data.encode('utf-8')
    return base64.urlsafe_b64encode(data).decode('utf-8').rstrip('=')


def decompose_jwt(jwt):
    """Decompose JWT for optimized transmission
    
    Args:
        jwt (str): Full JWT token
        
    Returns:
        dict: Dictionary with header, payload (raw JSON), and signature
    
    Operations: 1 base64 decode (payload only)
    """
    if not jwt:
        return None
    
    parts = jwt.split('.')
    if len(parts) != 3:
        logger.warning('Invalid JWT format')
        return None
    
    header_b64, payload_b64, signature_b64 = parts
    
    try:
        # Decode payload only - ONLY OPERATION
        payload_json = base64url_decode(payload_b64).decode('utf-8')
        
        result = {
            'header': header_b64,         # Keep as-is (supports IdPs with kid, jku, etc.)
            'payload': payload_json,      # Raw JSON (~25% smaller than base64)
            'signature': signature_b64    # Keep as-is
        }
        
        return result
        
    except Exception as err:
        logger.warning(f'Failed to decompose JWT: {err}')
        return None


def reassemble_jwt(metadata):
    """Reassemble JWT from header, raw JSON payload, and signature
    
    Args:
        metadata: gRPC metadata tuple list
        
    Returns:
        str|None: Reassembled JWT or None
    
    Operations: 1 base64 encode (payload only)
    """
    # Convert metadata to dict
    metadata_dict = {}
    for key, value in metadata:
        if isinstance(value, bytes):
            value = value.decode('utf-8')
        metadata_dict[key] = value
    
    # Check for compressed JWT components (new format)
    payload_header = metadata_dict.get('x-jwt-payload')
    signature = metadata_dict.get('x-jwt-sig')
    header_b64 = metadata_dict.get('x-jwt-header')
    
    if payload_header and signature and header_b64:
        try:
            # Base64url encode the raw JSON payload - ONLY OPERATION
            payload_b64 = base64url_encode(payload_header)
            
            # Reassemble JWT using original header
            jwt = f'{header_b64}.{payload_b64}.{signature}'
            
            return jwt
            
        except Exception as err:
            logger.warning(f'Failed to reassemble JWT: {err}')
            return None
    
    # Fall back to standard authorization header
    auth_header = metadata_dict.get('authorization')
    if auth_header and auth_header.startswith('Bearer '):
        jwt = auth_header[7:]
        return jwt
    
    return None


def add_compressed_jwt(metadata, jwt):
    """Add compressed JWT to metadata
    
    Args:
        metadata: gRPC metadata tuple list
        jwt (str): Full JWT token
    """
    if not jwt or not is_jwt_compression_enabled():
        # Fallback to standard authorization header
        metadata.append(('authorization', f'Bearer {jwt}'))
        return
    
    components = decompose_jwt(jwt)
    if not components:
        # Failed to decompose, use standard header
        metadata.append(('authorization', f'Bearer {jwt}'))
        return
    
    # Add compressed components: header + raw JSON payload + signature
    metadata.append(('x-jwt-header', components['header']))
    metadata.append(('x-jwt-payload', components['payload']))
    metadata.append(('x-jwt-sig', components['signature']))
